File: apps/scraper/query/sources/library_catalog.py
"""The ASU Library catalog: books, articles, journals and media."""

from __future__ import annotations

import urllib.parse

from scraper.core.types import QueryParam, QuerySource
from scraper.query.params import choices, code, text

_TYPES = {
    "all": "any",
    "books": "books",
    "articles": "articles",
    "journals": "journals",
    "images": "images",
    "scores": "scores",
    "maps": "maps",
    "sound recordings": "audios",
    "video": "videos",
}


def to_url(params: dict[str, str]) -> str:
    """The catalog search URL. Primo takes its query as a comma-separated expression."""
    query = {
        "query": f"any,contains,{text(params, 'keywords')},AND",
        "pfilter": f"rtype,exact,{code(params, 'type', _TYPES) or 'any'}",
        "tab": "LibraryCatalog",
        "search_scope": "MyInstitution",
        "vid": "01ASU_INST:01ASU",
        "lang": "en",
        "mode": "advanced",
        "offset": "0",
    }
    return "https://search.lib.asu.edu/discovery/search?" + urllib.parse.urlencode(query)


QUERY = QuerySource(
    key="library_catalog",
    description="Search the ASU Library catalog for books, articles, journals and media.",
    params=(
        QueryParam("keywords", "Title, author or subject.", required=True, example="deep learning"),
        QueryParam("type", "Resource type.", choices=choices(_TYPES)),
    ),
    to_url=to_url,
    needs_js=True,
    category="library",
)

File: apps/engine/src/runtime/tools/knowledge/search/library_catalog.rs
//! Books, articles, journals and media in the ASU Library.

use super::{LiveSource, Param};

/// The ASU Library catalog.
pub struct LibraryCatalog;

const TYPES: &[&str] = &[
    "all",
    "books",
    "articles",
    "journals",
    "images",
    "scores",
    "maps",
    "sound recordings",
    "video",
];

const PARAMS: &[Param] = &[
    Param::text("keywords", "Title, author or subject.")
        .required()
        .example("deep learning"),
    Param::one_of("type", "Resource type.", TYPES),
];

impl LiveSource for LibraryCatalog {
    fn key(&self) -> &'static str {
        "library_catalog"
    }

    fn hint(&self) -> &'static str {
        "books, articles and media"
    }

    fn category(&self) -> &'static str {
        "library"
    }

    fn params(&self) -> &'static [Param] {
        PARAMS
    }
}
